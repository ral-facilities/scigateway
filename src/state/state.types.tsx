import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
  ScheduledMaintenanceState,
  MaintenanceState,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  unauthorised?: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
  darkMode: boolean;
  highContrastMode: boolean;
  homepageUrl?: string;
  logo?: string;
  scheduledMaintenance: ScheduledMaintenanceState;
  maintenance: MaintenanceState;
  navigationDrawerLogo?: LogoState;
  adminPageDefaultTab?: 'maintenance' | 'download';
  contactUsAccessibilityFormUrl?: string;
  primaryColour?: string;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  isAdmin: boolean;
  avatarUrl: string;
}

export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  isAdmin: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  fetchScheduledMaintenanceState?: () => Promise<ScheduledMaintenanceState>;
  setScheduledMaintenanceState?: (
    scheduledMaintenanceState: ScheduledMaintenanceState
  ) => Promise<string | void>;
  fetchMaintenanceState?: () => Promise<MaintenanceState>;
  setMaintenanceState?: (
    maintenanceState: MaintenanceState
  ) => Promise<string | void>;
  redirectUrl: string | null;
  authUrl: string | undefined;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}

export interface LogoState {
  light: string;
  dark: string;
  altTxt: string;
}
