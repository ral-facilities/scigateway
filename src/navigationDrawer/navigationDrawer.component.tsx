import React, { Fragment, useCallback } from 'react';
import { Box, styled, Theme, Typography, useMediaQuery } from '@mui/material';
import Drawer from '@mui/material/Drawer';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import { useDispatch, useSelector } from 'react-redux';
import { Link, LinkProps } from 'react-router-dom';
import { PluginConfig } from '../state/scigateway.types';
import { StateType } from '../state/state.types';
import { structureMenuData } from '../state/pluginhelper';
import STFCLogoWhiteText from '../images/stfc-logo-white-text.png';
import STFCLogoBlueText from '../images/stfc-logo-blue-text.png';
import { getAppStrings, getString } from '../state/strings';
import { useTheme } from '@mui/material/styles';
import { toggleDrawer } from '../state/actions/scigateway.actions';

const LogoImage = styled('img')(({ theme }) => ({
  paddingRight: theme.spacing(2),
  paddingLeft: theme.spacing(2),
  width: '188px',
  paddingBottom: 24,
  color: theme.palette.text.secondary,
}));

// This has been adapted from the MaterialUI composition guide
// (https://material-ui.com/guides/composition/)
const ForwardRefLink = React.forwardRef<HTMLAnchorElement, LinkProps>(
  (linkProps, ref) => <Link innerRef={ref} {...linkProps} />
);
ForwardRefLink.displayName = 'ForwardRefLink';

export const NavigationDrawer = (): React.ReactElement => {
  const isDrawerOpen = useSelector(
    (state: StateType) => state.scigateway.drawerOpen
  );
  const plugins = useSelector((state: StateType) => state.scigateway.plugins);
  const isDarkMode = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  const homepageUrl = useSelector(
    (state: StateType) => state.scigateway.homepageUrl
  );
  const res = useSelector((state: StateType) =>
    getAppStrings(state, 'navigation-drawer')
  );
  const navigationDrawerLogo = useSelector(
    (state: StateType) => state.scigateway.navigationDrawerLogo
  );

  const dispatch = useDispatch();

  const theme = useTheme();
  const isViewportMdOrLarger = useMediaQuery(theme.breakpoints.up('md'));

  const createLink = useCallback(
    (plugin: PluginConfig, index: number): React.ReactElement => {
      const imgSrc = isDarkMode ? plugin.logoDarkMode : plugin.logoLightMode;

      const prefix = !imgSrc && plugin.logoAltText ? plugin.logoAltText : '';

      const displayText = plugin.displayName
        ? prefix + plugin.displayName
        : plugin.plugin;

      return (
        <ListItem
          key={index}
          component={ForwardRefLink}
          to={plugin.link}
          id={`plugin-link-${plugin.link.replace(/\//g, '-')}`}
          button
          dense
        >
          <ListItemText
            inset={!imgSrc}
            primary={displayText}
            primaryTypographyProps={{
              variant: 'subtitle1',
              sx: {
                textAlign: 'left',
                fontWeight: 'bold',
                color: (theme: Theme) => theme.colours.blue,
              },
            }}
          />
        </ListItem>
      );
    },
    [isDarkMode]
  );

  const buildMenuSection = useCallback(
    (
      sectionName: string,
      plugins: PluginConfig[],
      index: number
    ): React.ReactElement => {
      return (
        <Fragment key={index}>
          <Typography
            variant="h6"
            sx={{
              textAlign: 'left',
              paddingTop: 2,
              paddingBottom: 0,
              color: (theme: Theme) => theme.colours.contrastGrey,
              paddingLeft: 2,
            }}
          >
            {sectionName}
          </Typography>
          <List component="nav">
            {plugins.map((p, i) => {
              return p.link ? createLink(p, i) : null;
            })}
          </List>
        </Fragment>
      );
    },
    [createLink]
  );

  const renderRoutes = useCallback((): React.ReactFragment => {
    // don't include link to homepage in nav bar
    const filteredPlugins = homepageUrl
      ? plugins.filter((plugin) => plugin.link !== homepageUrl)
      : plugins;

    // Do not include admin plugins or plugins that explicitly ask to hide in the drawer list
    const sectionPlugins = structureMenuData(
      filteredPlugins.filter((plugin) => !plugin.admin && !plugin.hideFromMenu)
    );

    return (
      <Fragment>
        {Object.keys(sectionPlugins)
          .sort()
          .map((section, i) =>
            buildMenuSection(
              section,
              sectionPlugins[section] as PluginConfig[],
              i
            )
          )}
      </Fragment>
    );
  }, [buildMenuSection, plugins, homepageUrl]);

  const altTxt =
    navigationDrawerLogo?.altTxt ?? getString(res, 'alternative-text');

  // check if there is a custom logo supplied
  // if not, fallback to the default STFC logo
  let drawerLogoToUse;
  if (navigationDrawerLogo) {
    drawerLogoToUse = isDarkMode
      ? navigationDrawerLogo.dark
      : navigationDrawerLogo.light;
  } else {
    drawerLogoToUse = isDarkMode ? STFCLogoWhiteText : STFCLogoBlueText;
  }

  return (
    <Drawer
      sx={{
        width: (theme: Theme) => theme.drawerWidth,
        flexShrink: 0,
      }}
      variant={isViewportMdOrLarger ? 'persistent' : undefined}
      anchor="left"
      open={isDrawerOpen}
      onClose={() => dispatch(toggleDrawer())}
      PaperProps={
        isViewportMdOrLarger
          ? {
              sx: (theme: Theme) => ({
                width: theme.drawerWidth,
                background: theme.palette.background.default,
                top: theme.mainAppBarHeight,
                height: `calc(100% - ${theme.footerPaddingBottom} - ${theme.footerPaddingTop} - ${theme.footerHeight} - ${theme.mainAppBarHeight})`,
                position: 'absolute',
              }),
            }
          : {}
      }
    >
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="flex-start"
        height="100%"
        boxSizing="border-box"
      >
        {renderRoutes()}

        {drawerLogoToUse && (
          <Box marginTop="auto">
            <LogoImage alt={altTxt} src={drawerLogoToUse} />
          </Box>
        )}
      </Box>
    </Drawer>
  );
};

export default NavigationDrawer;
