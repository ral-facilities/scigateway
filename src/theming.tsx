import {
  createTheme,
  ThemeProvider,
  StyledEngineProvider,
  Theme,
} from '@mui/material/styles';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';
import { checkboxClasses } from '@mui/material/Checkbox';
import {
  formHelperTextClasses,
  formLabelClasses,
  inputClasses,
  outlinedInputClasses,
  tabClasses,
} from '@mui/material';

/* UKRI colours */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const UKRI_COLOURS = {
  blue: '#2E2D62',
  bright: {
    orange: '#FF6900', // pure orange
    yellow: '#FBBB10', // yellow
    green: '#67C04D', // light green
    blue: '#1E5DF8', // blue
    purple: '#BE2BBB', // bright purple
    red: '#E94D36', // light red
  },
  deep: {
    orange: '#C13D33', // pure orange
    yellow: '#F08900', // vivid yellow
    green: '#3E863E', // green
    blue: '#003088', // blue
    purple: '#8A1A9B', // bright purple
    red: '#A91B2E', // red
  },
};

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#C34F00',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  /* The type of mode this is */
  type: 'default' | 'contrast';

  /* Primary/secondary colours used for MUI */
  primary: string;
  secondary: string;

  /* Secondary text colour */
  textSecondary: string;

  /* Background colours for the page and papers */
  background: string;
  paper: string;

  /* Standard colours used in plugins (change to lighter/darker shades
    between light and dark modes) - these are meant to give good contrast
    for text on the chosen paper background colour */
  blue: string;
  orange: string;
  red: string;
  grey: string;

  /* Lighter colours */
  lightBlue: string; //Used for lighter coloured buttons
  lightOrange: string; //Used for notifcation icon

  /* Darker colours */
  darkGreen: string; //Used for cookie consent message
  darkOrange: string; //Used for help tour

  /* Contrast colours that need to change significantly between dark
     and light modes */
  contrastGrey: string; //Used on homepage

  /* Informational/Warning colours */
  information: string; //Used in open data label
  warning: string; //Used for selection alert banner

  /* Colours for <a> style links */
  link: {
    default: string;
    visited: string;
    active: string;
  };

  footerLink: {
    default: string;
    active: string;
  };

  chip: string; //Used for chip colours in cards and filters

  /* Colours used on the homepage */
  homePage: {
    heading: string;
    blueDescription: string;
    blueButton: string;
  };

  /* Colour for Tabs in search */
  tabsGrey: string;
}

declare module '@mui/material/styles' {
  interface Theme {
    colours: ThemeColours;
    drawerWidth: string;
    footerPaddingTop: string;
    footerPaddingBottom: string;
    footerHeight: string;
    mainAppBarHeight: string;
  }
  // allow configuration using `createTheme`
  interface ThemeOptions {
    colours: ThemeColours;
    drawerWidth: string;
    footerPaddingTop: string;
    footerPaddingBottom: string;
    footerHeight: string;
    mainAppBarHeight: string;
  }
}

const DARK_MODE_COLOURS: ThemeColours = {
  type: 'default',
  primary: UKRI_COLOURS.deep.blue,
  secondary: '#80ACFF',
  textSecondary: 'rgba(255, 255, 255, 0.7)',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#F26300',
  red: '#FF7F73',
  grey: '#A4A4A4',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#E6E6E6',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#86B4FF',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#595959',
  homePage: {
    heading: '#FFFFFF',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#3A3A3A',
};

const DARK_MODE_HIGH_CONTRAST_COLOURS: ThemeColours = {
  type: 'contrast',
  primary: '#86B4FF',
  secondary: '#80ACFF',
  textSecondary: 'rgba(255, 255, 255, 0.7)',
  background: '#000000',
  paper: '#1A1A1A',
  blue: '#B4CCFA',
  orange: '#FFC14D',
  red: '#FFA198',
  grey: '#A4A4A4',
  lightBlue: UKRI_COLOURS.deep.blue,
  lightOrange: '#FFC14D',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#FFFFFF',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFC14D',
  link: {
    default: '#B4CCFA',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#000000',
    active: '#851D0F',
  },
  chip: '#3A3A3A',
  homePage: {
    heading: '#FFFFFF',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#1A1A1A',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  type: 'default',
  primary: UKRI_COLOURS.deep.blue,
  secondary: UKRI_COLOURS.deep.blue,
  textSecondary: 'rgba(0, 0, 0, 0.54)',
  background: '#FAFAFA',
  paper: '#FFFFFF',
  blue: UKRI_COLOURS.deep.blue,
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#717171',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#1E5DF8',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#E0E0E0',
  homePage: {
    heading: '#333333',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#EEEEEE',
};

const LIGHT_MODE_HIGH_CONTRAST_COLOURS: ThemeColours = {
  type: 'contrast',
  primary: UKRI_COLOURS.deep.blue,
  secondary: UKRI_COLOURS.deep.blue,
  textSecondary: '#000000',
  background: '#FAFAFA',
  paper: '#FFFFFF',
  blue: '#002466',
  orange: '#C34F00',
  red: '#801100',
  grey: '#727272',
  lightBlue: UKRI_COLOURS.bright.blue,
  lightOrange: '#FF6900',
  darkGreen: '#3A7F3A',
  darkOrange: STATIC_COLOURS.orange,
  contrastGrey: '#000000',
  information: UKRI_COLOURS.deep.blue,
  warning: '#FFA500',
  link: {
    default: '#052d94',
    visited: '#BE2BBB',
    active: '#E94D36',
  },
  footerLink: {
    default: '#FFFFFF',
    active: '#E58885',
  },
  chip: '#E0E0E0',
  homePage: {
    heading: '#000000',
    blueDescription: '#FFFFFF',
    blueButton: UKRI_COLOURS.bright.blue,
  },
  tabsGrey: '#EEEEEE',
};

export const buildTheme = (
  darkModePreference: boolean,
  highContrastModePreference?: boolean,
  primaryColour?: string
): Theme => {
  const colours = darkModePreference
    ? highContrastModePreference
      ? DARK_MODE_HIGH_CONTRAST_COLOURS
      : DARK_MODE_COLOURS
    : highContrastModePreference
      ? LIGHT_MODE_HIGH_CONTRAST_COLOURS
      : LIGHT_MODE_COLOURS;

  if (primaryColour && !(darkModePreference && highContrastModePreference)) {
    colours.primary = primaryColour;
  }

  const componentOverrides = {
    MuiPaper: {
      styleOverrides: { root: { backgroundImage: 'unset' } },
    },
    MuiLink: {
      defaultProps: {
        underline: 'hover' as 'always' | 'none' | 'hover' | undefined,
      },
      styleOverrides: {
        root: {
          color: colours.blue,
        },
      },
    },
    MuiTabs: {
      styleOverrides: {
        indicator: {
          color: STATIC_COLOURS.darkBlue,
          textDecoration: 'underline',
        },
      },
    },
    MuiTab: {
      styleOverrides: {
        textColorSecondary: {
          [`&.${tabClasses.selected}`]: {
            color: darkModePreference ? '#FFFFFF' : colours.blue,
          },
        },
      },
    },
    MuiCheckbox: {
      styleOverrides: {
        colorPrimary: {
          [`&.${checkboxClasses.checked}`]: {
            color: colours.blue,
          },
          [`&.${checkboxClasses.indeterminate}`]: {
            color: colours.grey,
          },
        },
      },
    },
    MuiFormLabel: {
      styleOverrides: {
        root: {
          [`&.${formLabelClasses.error}`]: {
            color: colours.red,
          },
          [`&.${formLabelClasses.focused}`]: {
            color: colours.blue,
          },
        },
        asterisk: {
          [`&.${formLabelClasses.error}`]: {
            color: colours.red,
          },
        },
      },
    },
    MuiBadge: {
      styleOverrides: {
        colorPrimary: {
          backgroundColor: STATIC_COLOURS.orange,
          color: 'white',
        },
      },
    },
    MuiInput: {
      styleOverrides: {
        underline: {
          [`&.${inputClasses.error}`]: {
            borderBottomColor: colours.red,
          },
          [`&.${inputClasses.error}:after`]: {
            borderBottomColor: colours.blue,
          },
        },
      },
    },
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
          [`&.${outlinedInputClasses.error} .${outlinedInputClasses.notchedOutline}`]:
            {
              borderColor: colours.red,
            },
        },
        //Override opacity of placeholder text
        input: {
          '&::placeholder': {
            colour: colours.textSecondary,
            opacity:
              !darkModePreference && highContrastModePreference ? 1.0 : 0.7,
          },
        },
      },
    },
    MuiFormHelperText: {
      styleOverrides: {
        root: {
          [`&.${formHelperTextClasses.error}`]: {
            color: colours.red,
          },
        },
      },
    },
    MuiChip: {
      styleOverrides: {
        root: {
          backgroundColor: colours.chip,
        },
      },
    },
    MuiPickersToolbar: {
      styleOverrides: {
        toolbar: {
          backgroundColor: colours.primary,
        },
      },
    },
    MuiPickersCalendarHeader: {
      styleOverrides: {
        dayLabel: {
          color: colours.grey,
        },
      },
    },
    MuiPickersDay: {
      styleOverrides: {
        current: {
          color: colours.blue,
        },
        dayDisabled: {
          color: colours.grey,
        },
      },
    },
    MuiPickersYear: {
      styleOverrides: {
        root: {
          '&:active': {
            color: colours.blue,
          },
          '&:focus': {
            color: colours.blue,
          },
        },
        yearSelected: {
          color: colours.blue,
        },
        yearDisabled: {
          color: colours.grey,
        },
      },
    },
    MuiPickersMonth: {
      styleOverrides: {
        root: {
          '&:active': {
            color: colours.blue,
          },
          '&:focus': {
            color: colours.blue,
          },
        },
        monthSelected: {
          color: colours.blue,
        },
        monthDisabled: {
          color: colours.grey,
        },
      },
    },
  };

  return createTheme({
    palette: {
      // Light/dark mode
      mode: darkModePreference ? 'dark' : 'light',
      primary: {
        main: colours.primary,
      },
      secondary: {
        main: colours.secondary,
      },
      text: {
        secondary: colours.textSecondary,
      },
      background: {
        default: colours.background,
        paper: colours.paper,
      },
    },
    typography: {
      button: {
        textTransform: 'none',
      },
    },
    components: componentOverrides,
    colours: colours,
    drawerWidth: '220px',
    footerPaddingTop: '8px',
    footerPaddingBottom: '8px',
    footerHeight: '20px',
    mainAppBarHeight: '64px',
  });
};

function mapThemeProviderStateToProps(state: StateType): {
  prefersDarkMode: boolean;
} {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  const highContrastModePreference = useSelector(
    (state: StateType) => state.scigateway.highContrastMode
  );
  const primaryColour = useSelector(
    (state: StateType) => state.scigateway.primaryColour
  );
  return (
    <StyledEngineProvider injectFirst>
      <ThemeProvider
        theme={buildTheme(
          darkModePreference,
          highContrastModePreference,
          primaryColour
        )}
      >
        {props.children}
      </ThemeProvider>
    </StyledEngineProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
