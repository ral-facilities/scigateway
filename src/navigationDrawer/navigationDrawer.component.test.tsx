import * as React from 'react';
import { NavigationDrawer } from './navigationDrawer.component';

import { PluginConfig } from '../state/scigateway.types';
import {
  StyledEngineProvider,
  ThemeProvider,
  useMediaQuery,
} from '@mui/material';
import { MemoryRouter } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';
import { buildTheme } from '../theming';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { combineReducers, createStore, Store } from 'redux';
import { ScigatewayState, StateType } from '../state/state.types';
import ScigatewayReducer, {
  initialState as scigatewayInitialState,
} from '../state/reducers/scigateway.reducer';

jest.mock('@mui/material', () => ({
  __esmodule: true,
  ...jest.requireActual('@mui/material'),
  useMediaQuery: jest.fn(),
}));

describe('Navigation drawer component', () => {
  let history: History;
  const theme = buildTheme(false);

  beforeEach(() => {
    history = createMemoryHistory();
    history.replace('/help');
    // I don't think MediaQuery works properly in jest
    // in the implementation useMediaQuery is used to query whether the current viewport is md or larger
    // here we assume it is always the case.
    jest.mocked(useMediaQuery).mockReturnValue(true);
  });

  function Wrapper({ children }: { children: React.ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </StyledEngineProvider>
      </MemoryRouter>
    );
  }

  function createMockStore(initialState: Partial<ScigatewayState>): Store {
    return createStore(
      combineReducers<Partial<StateType>>({
        scigateway: (
          state = { ...scigatewayInitialState, ...initialState },
          action
        ) => ScigatewayReducer(state, action),
      })
    );
  }

  it('Navigation drawer renders correctly when open', () => {
    const { asFragment } = render(
      <Provider
        store={createMockStore({
          drawerOpen: true,
          plugins: [],
          darkMode: false,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Navigation drawer renders correctly when closed', () => {
    const { asFragment } = render(
      <Provider
        store={createMockStore({
          drawerOpen: false,
          plugins: [],
          darkMode: false,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  function buildPlugin(
    order: number,
    displayName: string,
    section: string
  ): PluginConfig {
    return {
      order: order,
      displayName: displayName,
      section: section,
      link: 'link',
      plugin: 'plugin',
    };
  }

  it('renders a plugin list grouped by sections ordered alphabetically when open', () => {
    const dummyPlugins: PluginConfig[] = [
      buildPlugin(1, 'data-plugin', 'DATA'),
      buildPlugin(2, 'data-plugin2', 'DATA'),
      buildPlugin(1, 'analysis-plugin2', 'ANALYSIS'),
      buildPlugin(2, 'data-plugin1', 'DATA'),
      buildPlugin(-1, 'analysis-plugin', 'ANALYSIS'),
      {
        order: 3,
        plugin: 'data-plugin-no-displayname',
        link: 'plugin_link',
        section: 'DATA',
      },
    ];

    render(
      <Provider
        store={createMockStore({
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: false,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    const navBarLinks = screen.getAllByRole('link');
    expect(navBarLinks).toHaveLength(6);

    expect(navBarLinks[0]).toHaveTextContent('analysis-plugin');
    expect(navBarLinks[1]).toHaveTextContent('analysis-plugin2');
    expect(navBarLinks[2]).toHaveTextContent('data-plugin');
    expect(navBarLinks[3]).toHaveTextContent('data-plugin1');
    expect(navBarLinks[4]).toHaveTextContent('data-plugin2');
    expect(navBarLinks[5]).toHaveTextContent('data-plugin-no-displayname');
  });

  it('does not render admin plugins or plugins that ask to hide in list', () => {
    const dummyPlugins: PluginConfig[] = [
      {
        order: 0,
        plugin: 'data-plugin',
        link: 'plugin_link',
        section: 'DATA',
        displayName: 'display name',
        admin: true,
      },
      {
        order: 1,
        plugin: 'data-plugin-2',
        link: 'plugin_link_2',
        section: 'DATA',
        displayName: 'display name 2',
        hideFromMenu: true,
      },
    ];

    render(
      <Provider
        store={createMockStore({
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: false,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(screen.queryAllByRole('link')).toHaveLength(0);
  });

  it('does not display link to homepage if a homepage link is set', () => {
    const homepageLink = 'homepage';
    const dummyPlugins: PluginConfig[] = [
      {
        order: 0,
        plugin: 'homepage-plugin',
        link: homepageLink,
        section: 'Homepage',
        displayName: 'home page',
      },
      {
        order: 1,
        plugin: 'data-plugin-no-displayname',
        link: 'plugin_link',
        section: 'DATA',
        displayName: 'display name',
      },
    ];

    render(
      <Provider
        store={createMockStore({
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: false,
          res: undefined,
          homepageUrl: homepageLink,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(screen.queryByRole('link', { name: 'home page' })).toBeNull();
  });

  it('renders the dark mode logo at the top', () => {
    const dummyPlugins: PluginConfig[] = [
      {
        order: 0,
        plugin: 'data-plugin',
        link: 'plugin_link',
        section: 'DATA',
        displayName: '\xa0display name',
      },
    ];

    render(
      <Provider
        store={createMockStore({
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: true,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(screen.getByRole('img')).toHaveAttribute(
      'src',
      'stfc-logo-white-text.png'
    );
  });

  it('should be able to use logo set in the settings file', () => {
    const dummyPlugins: PluginConfig[] = [
      {
        order: 0,
        plugin: 'data-plugin',
        link: 'plugin_link',
        section: 'DATA',
        displayName: '\xa0display name',
      },
    ];
    const navigationDrawerLogo = {
      light: '/test/lightmode',
      dark: '/test/darkmode',
      altTxt: 'alt txt test',
    };

    const { rerender } = render(
      <Provider
        store={createMockStore({
          navigationDrawerLogo,
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: false,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>,
      { wrapper: Wrapper }
    );

    expect(screen.getByAltText('alt txt test')).toHaveAttribute(
      'src',
      '/test/lightmode'
    );

    rerender(
      <Provider
        store={createMockStore({
          navigationDrawerLogo,
          drawerOpen: true,
          plugins: dummyPlugins,
          darkMode: true,
          res: undefined,
        })}
      >
        <NavigationDrawer />
      </Provider>
    );

    expect(screen.getByAltText('alt txt test')).toHaveAttribute(
      'src',
      '/test/darkmode'
    );
  });
});
