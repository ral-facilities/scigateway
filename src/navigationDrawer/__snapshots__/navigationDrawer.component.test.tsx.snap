// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation drawer component Navigation drawer renders correctly when closed 1`] = `
<DocumentFragment>
  <div
    class="MuiDrawer-root MuiDrawer-docked css-ak80xd-MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-8bwi45-MuiPaper-root-MuiDrawer-paper"
      style="visibility: hidden; webkit-transform: translateX(-0px); transform: translateX(-0px);"
    >
      <div
        class="MuiBox-root css-1pvs5pb"
      >
        <div
          class="MuiBox-root css-8xl60i"
        >
          <img
            alt="alternative-text"
            class="css-1taxr2z"
            src="stfc-logo-blue-text.png"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navigation drawer component Navigation drawer renders correctly when open 1`] = `
<DocumentFragment>
  <div
    class="MuiDrawer-root MuiDrawer-docked css-ak80xd-MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-8bwi45-MuiPaper-root-MuiDrawer-paper"
    >
      <div
        class="MuiBox-root css-1pvs5pb"
      >
        <div
          class="MuiBox-root css-8xl60i"
        >
          <img
            alt="alternative-text"
            class="css-1taxr2z"
            src="stfc-logo-blue-text.png"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
